Add Package version details and approval fields

praveenkuttappan opened the pull request:
- Added package name, package version and approval fields in review file and revision models.
- Package name and package version is set from C# language processors.

praveenkuttappan commented:
@JonathanGiles : FYI.. This PR has changes to add language and other fields in review

JonathanGiles commented:
@praveenkuttappan I note that you've updated the model version to 19, so I will do that for Java too. I'm curious if you need me to add support for approval fields in the json, and if so, what fields and what is expected?

praveenkuttappan commented:
@JonathanGiles  Approval fields are not required in language specific code files. It's stored in revision metadata in cosmos DB.

pakrym commented:
Hm, I'm not a fan of adding these properties without having any consumers for them. Let's instead add them as part of implementing the features they would be a part of.

praveenkuttappan commented:
> Hm, I'm not a fan of adding these properties without having any consumers for them. Let's instead add them as part of implementing the features they would be a part of.

I got it. Reason for adding these properties is to start getting these properties updated by language processors. My plan was implement them in chunks rather than a a big functionality. I will check and see if I can pull in more changes that uses these properties.

pakrym commented:
>Reason for adding these properties is to start getting these properties updated by language processors

I'm OK with this one if we limit it to properties actually being populated like language/version etc. But we also need to decide what some of them mean across languages.

praveenkuttappan commented:
> > Reason for adding these properties is to start getting these properties updated by language processors
> 
> I'm OK with this one if we limit it to properties actually being populated like language/version etc. But we also need to decide what some of them mean across languages.
> 
> Approval related properties are not generator-provided.

I have reverted changes and also removed  packageVersion field as we discussed.

praveenkuttappan commented:
@JonathanGiles : Can you please remove PackageVersion from Codefile in java code generator. Only PackageName needs to be filled by JAva processor. Language is anyway added by APIView tool when we process review file from java parser.

praveenkuttappan commented:
@pakrym @JonathanGiles : Here is the first version of Language filter UI change as part of this PR. 
![image](https://user-images.githubusercontent.com/55455725/98914818-f67f9380-247d-11eb-826d-715f91f313a2.png)

JonathanGiles commented:
I'm not clear on what you want me to support. At present I have added language, package name and package version. Do you want me to just disable package version for now and add it back later, or do you not think you need it now? You still need language, correct?

praveenkuttappan commented:
> I'm not clear on what you want me to support. At present I have added language, package name and package version. Do you want me to just disable package version for now and add it back later, or do you not think you need it now? You still need language, correct?

You can just remove PackageVersion field. We don’t need it at the moment.

pakrym commented on src/dotnet/APIView/APIViewWeb/Repositories/CosmosReviewRepository.cs:
```diff
@@ -33,6 +33,10 @@ public async Task<IEnumerable<ReviewModel>> GetReviewsAsync(bool closed)
                 allReviews.AddRange(result.Resource);
             }
 
+            if(!string.IsNullOrEmpty(Language) && Language != "All")
```
we should filter this as part of the query same as we do with IsClosed.

praveenkuttappan replied:
Language is not saved in Review model currently. All of existing records has language in revisionmodel which is child of review. But you are right. We need packagename in ReviewModel itself to query and find current master version is approved or not.

pakrym replied:
I think we can use the query for inner objects as well. Cosmos query language should allow it.

praveenkuttappan replied:
I have fixed this to refer inner variables using Exists clause

pakrym commented on src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:
```diff
@@ -60,9 +60,9 @@ public async Task<ReviewModel> CreateReviewAsync(ClaimsPrincipal user, string or
         }
 
-        public Task<IEnumerable<ReviewModel>> GetReviewsAsync(bool closed)
+        public Task<IEnumerable<ReviewModel>> GetReviewsAsync(bool closed, string Language)
```
```suggestion
        public Task<IEnumerable<ReviewModel>> GetReviewsAsync(bool closed, string language)
```

pakrym commented on src/dotnet/APIView/APIViewWeb/Repositories/ReviewManager.cs:
```diff
@@ -255,6 +255,7 @@ private void InitializeFromCodeFile(ReviewCodeFileModel file, CodeFile codeFile)
             file.Language = codeFile.Language;
             file.VersionString = codeFile.VersionString;
             file.Name = codeFile.Name;
```
How is name different from the package name?

praveenkuttappan replied:
Currently Name is package name and version name. Package Name is to standardize this to have only package name for lookup purpose when autodetecting API difference for a package is added.

pakrym commented on src/dotnet/APIView/APIViewWeb/Models/ReviewModel.cs:
```diff
@@ -58,7 +58,7 @@ public bool IsUserSubscribed(ClaimsPrincipal user)
         public string GetUserEmail(ClaimsPrincipal user) =>
             NotificationManager.GetUserEmail(user);
-
+        
```
nit: ws

pakrym commented on src/dotnet/APIView/APIViewWeb/Models/ReviewModel.cs:
```diff
@@ -75,6 +75,9 @@ public string DisplayName
 
         [JsonIgnore]
-        public string Language => Revisions.FirstOrDefault()?.Files.FirstOrDefault()?.Language;
+        public string Language => Revisions.LastOrDefault()?.Files.LastOrDefault()?.Language;
+
+        [JsonIgnore]
+        public string PackageName => Revisions.LastOrDefault()?.Files.LastOrDefault()?.PackageName;
```
Are these two used?

praveenkuttappan replied:
Reverted this change. Not required anymore in current PR.

pakrym commented:
Can you deploy to staging for a final look please?

praveenkuttappan commented:
This change is deployed on staging instance now. Please note that there are few json reviews incorrectly tagged as C# which should get fixed when they upload new version of wheel going forward.

pakrym reviewed (approved):
Great job!
